### PR TITLE
Add graphy

### PR DESCRIPTION
**Pull request by aarondl**

### CODE REVIEW

**Comment by aarondl** [comment]
Sample output:
```make-compile::restore::clear::4e255efa754d91b825476b57e111345f200944e1.tar.xz,0  
make-compile::restore::extract::90bf9699731d951034bb08eb850190872e8e4890.tar.xz,6000000000
create-release::docker::hcf-versions,15000000000                                 
create-release::docker::cf-open-autoscaler,8000000000                            
make-compile::restore::clear::e19239db58fb91edb5f670e24ef793d266fd6544.tar.xz,0  
make-compile::restore::extract::91cf17318eb58c2a4840e7e4615453f6c78cee3c.tar.xz,0,0
make-compile::restore::extract::b1c360043f27223e0cf4b8a53bbe75ac5f1faf87.tar.xz,1000000000```

File: cmd/graphy/main.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	times := make([]string, len(headers))
+	for i := 0; ; i++ {
+		foundTiming := false
+		for j := 0; j < len(headers); j++ {
+			timings := events[headers[j]]
+			if i < len(timings) {
+				times[j] = strconv.Itoa(int(timings[i]))
+				foundTiming = true
+			} else {
+				times[j] = "0"
```
Comment (mook-as) [inline cmd/graphy/main.go]: Consider leaving times with no value empty (as opposed to `0`) to make it possible to distinguish between "no value" and "so fast we couldn't measure it".

File: cmd/graphy/main.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	writer := csv.NewWriter(os.Stdout)
+	var headers []string
+	for series := range events {
+		headers = append(headers, series)
+	}
+	if err := writer.Write(headers); err != nil {
+		return fmt.Errorf("failed to write row header: %v", err)
+	}
+
+	times := make([]string, len(headers))
```
Comment (mook-as) [inline cmd/graphy/main.go]: I feel like this chunk (lines 109 ~ 130) could use a comment as a higher-level description of what it's doing. Maybe not. Not sure.

File: cmd/graphy/main.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	"os"
+	"strconv"
+	"time"
+)
+
+var usage = `Consumes a csv file in the format: YYYY-MM-DD HH:MM:SS,file,timeseries,(start|done)
+and outputs a new csv to stdout that aggregates the time differences between start/end
+
+Usage:
+  graphy [flags] <csvfilename>
```
Comment (aarondl) [inline cmd/graphy/main.go]: It's intention was terminal graphs when I first made it. But then I remembered Excel will do that better. Maybe one day it'll graph :p

File: cmd/graphy/main.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"time"
+)
+
+var usage = `Consumes a csv file in the format: YYYY-MM-DD HH:MM:SS,file,timeseries,(start|done)
+and outputs a new csv to stdout that aggregates the time differences between start/end
```
Comment (aarondl) [inline cmd/graphy/main.go]: Done.

File: cmd/graphy/main.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func help() {
+	fmt.Fprintln(os.Stderr, usage)
+}
+
+func graphyCmd(args []string) error {
+	if len(args) < 1 {
+		help()
+		return errors.New("need more arguments")
```
Comment (aarondl) [inline cmd/graphy/main.go]: Done because it was so easy.

File: cmd/graphy/main.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+	events := make(map[string][]uint64)
+	starts := make(map[string]time.Time)
+
+	reader := csv.NewReader(f)
+	for {
+		row, err := reader.Read()
+		if err == io.EOF {
+			break
+		} else if err != nil {
```
Comment (aarondl) [inline cmd/graphy/main.go]: `else if` feels cleaner here. Won't change without good reason. Takes more lines to write the alternative when in reality this block is "checking what err" is so makes sense to me as a single "piece". Could be a switch if you wanna be weird about it.

File: cmd/graphy/main.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+			return fmt.Errorf("failed to read row: %v", err)
```
Comment (aarondl) [inline cmd/graphy/main.go]: Done

File: cmd/graphy/main.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		}
+
+		date, err := time.Parse("2006-01-02 15:04:05", row[0])
+		if err != nil {
+			return fmt.Errorf("failed to parse time value: %v", err)
```
Comment (aarondl) [inline cmd/graphy/main.go]: Done

File: cmd/graphy/main.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+		}
+		series := row[2]
+		event := row[3]
+		if event == "start" {
+			starts[series] = date
+		} else if event == "done" {
+			startDate, ok := starts[series]
+			if !ok {
+				return fmt.Errorf("found done event with no start event: %s", series)
```
Comment (aarondl) [inline cmd/graphy/main.go]: Done

File: cmd/graphy/main.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+			}
+			events[series] = append(events[series], uint64(date.Sub(startDate)))
+			delete(starts, series)
+		}
```
Comment (aarondl) [inline cmd/graphy/main.go]: Rectified with comment showing intent to ignore

File: cmd/graphy/main.go (hunk 2 above)
Comment (aarondl) [inline cmd/graphy/main.go]: Renamed the functions to make it more clear. Also added a comment about the general shape of the algorithm.

File: cmd/graphy/main.go (hunk 5 above)
Comment (mook-as) [inline cmd/graphy/main.go]: Ah, I was thinking reading from stdin if there were no arguments, but an explicit `-` is good too.

File: cmd/graphy/main.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+		if !foundTiming {
+			break
+		}
+
+		if err := writer.Write(times); err != nil {
+			return fmt.Errorf("failed to write row: %v", err)
+		}
+	}
+	writer.Flush()
```
Comment (mook-as) [inline cmd/graphy/main.go]: Flushing might trigger an error (which you'd need to read separately).

**Comment by mook-as** [comment]
stampy

**Pull request merged**